File: backend/app/models/earthquake.py
from datetime import datetime

from sqlalchemy import (
    CheckConstraint,
    DateTime,
    Double,
    FetchedValue,
    Integer,
    String,
    Text,
    UniqueConstraint,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column

from app.database.base import Base


class Earthquake(Base):
    __tablename__ = "earthquakes"
    __table_args__ = (
        UniqueConstraint(
            "provider", "provider_event_id", name="uq_earthquakes_provider_event"
        ),
        CheckConstraint(
            "latitude >= -90 AND latitude <= 90", name="ck_earthquakes_latitude"
        ),
        CheckConstraint(
            "longitude >= -180 AND longitude <= 180",
            name="ck_earthquakes_longitude",
        ),
        CheckConstraint("version > 0", name="ck_earthquakes_version_positive"),
    )

    id: Mapped[str] = mapped_column(String(255), primary_key=True)
    provider: Mapped[str] = mapped_column(String(32), nullable=False)
    provider_event_id: Mapped[str] = mapped_column(String(255), nullable=False)
    kind: Mapped[str] = mapped_column(String(64), nullable=False)
    title: Mapped[str] = mapped_column(String(500), nullable=False)
    place: Mapped[str] = mapped_column(String(500), nullable=False)
    magnitude: Mapped[float] = mapped_column(Double, nullable=False)
    depth_km: Mapped[float] = mapped_column(Double, nullable=False)
    latitude: Mapped[float] = mapped_column(Double, nullable=False)
    longitude: Mapped[float] = mapped_column(Double, nullable=False)
    event_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, index=True
    )
    provider_updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, index=True
    )
    retrieved_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )
    review_status: Mapped[str | None] = mapped_column(String(64), nullable=True)
    source_url: Mapped[str] = mapped_column(Text, nullable=False)
    version: Mapped[int] = mapped_column(Integer, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        server_onupdate=FetchedValue(),
    )

File: mobile/lib/features/alerts/domain/earthquake.dart
enum AlertDataStatus { current, stale }

final class Earthquake {
  Earthquake({
    required this.id,
    required this.provider,
    required this.providerEventId,
    required this.kind,
    required this.title,
    required this.place,
    required this.magnitude,
    required this.depthKm,
    required this.latitude,
    required this.longitude,
    required this.eventAt,
    required this.providerUpdatedAt,
    required this.retrievedAt,
    required this.reviewStatus,
    required this.sourceUrl,
    required this.version,
  }) : assert(id.startsWith('usgs:')),
       assert(provider == 'usgs'),
       assert(kind == 'earthquake_information'),
       assert(sourceUrl.startsWith('https://')),
       assert(magnitude.isFinite),
       assert(depthKm.isFinite),
       assert(latitude.isFinite && latitude >= -90 && latitude <= 90),
       assert(longitude.isFinite && longitude >= -180 && longitude <= 180),
       assert(eventAt.isUtc),
       assert(providerUpdatedAt.isUtc),
       assert(retrievedAt.isUtc),
       assert(version > 0);

  final String id;
  final String provider;
  final String providerEventId;
  final String kind;
  final String title;
  final String place;
  final double magnitude;
  final double depthKm;
  final double latitude;
  final double longitude;
  final DateTime eventAt;
  final DateTime providerUpdatedAt;
  final DateTime retrievedAt;
  final String? reviewStatus;
  final String sourceUrl;
  final int version;

  @override
  bool operator ==(Object other) {
    return identical(this, other) ||
        other is Earthquake &&
            id == other.id &&
            provider == other.provider &&
            providerEventId == other.providerEventId &&
            kind == other.kind &&
            title == other.title &&
            place == other.place &&
            magnitude == other.magnitude &&
            depthKm == other.depthKm &&
            latitude == other.latitude &&
            longitude == other.longitude &&
            eventAt == other.eventAt &&
            providerUpdatedAt == other.providerUpdatedAt &&
            retrievedAt == other.retrievedAt &&
            reviewStatus == other.reviewStatus &&
            sourceUrl == other.sourceUrl &&
            version == other.version;
  }

  @override
  int get hashCode => Object.hash(
    id,
    provider,
    providerEventId,
    kind,
    title,
    place,
    magnitude,
    depthKm,
    latitude,
    longitude,
    eventAt,
    providerUpdatedAt,
    retrievedAt,
    reviewStatus,
    sourceUrl,
    version,
  );
}

final class AlertSnapshot {
  AlertSnapshot({
    required List<Earthquake> items,
    required this.dataStatus,
    required this.lastSuccessfulRefreshAt,
  }) : assert(lastSuccessfulRefreshAt.isUtc),
       items = List.unmodifiable(items);

  final List<Earthquake> items;
  final AlertDataStatus dataStatus;
  final DateTime lastSuccessfulRefreshAt;
}
